### ISSUE TITLE
Using private quarks in kernel functions

### ISSUE THREAD

**Issue by PhelypeOleinik** [latex3/latex3#593]
From here: https://tex.stackexchange.com/q/496780/134574  
I think that much like in the way `\l_tmpa_tl` is not used in the kernel to avoid undesired interaction with user code, the quarks used in kernel code should probably be internal to each kernel module as well. Perhaps provide a `\q__clist_mark`, `\q__clist_stop`, etc. for `l3clist`, and so on. Or is there a reason to use the same public quarks in the kernel code?

**Comment by wspr** [comment]
Yes, I agree — this part of the code was developed before the idea of private variables/functions was even thought of.

I'd go so far as to suggest that `\quark_new:N` should be a constructor for other functions like `\quark_if_<quark>:NTF`, `\use_none_to_<quark>:wN`, and so on — this would make it much easier for 3rd party quark users to safely integrate quarks into their own code.

**Comment by josephwright** [comment]
The `clist` data type is meant for 'simple` translation from user input: quarks seem out-of-place there. Isn't this why we have `seq`? In particular, the recursion quarks are only ever meant to be used during a mapping or similar, so there really should be no sensible case where they get stored.

**Comment by josephwright** [comment]
@blefloch is likely best-placed to comment here: is there a link to `\tl_trim_spaces:n`?

**Comment by FrankMittelbach** [comment]
> Yes, I agree — this part of the code was developed before the idea of private variables/functions was even thought of.

agreed --- even if they shouldn't appear in such places it might be best to have private quarks

> I'd go so far as to suggest that `\quark_new:N` should be a constructor for other functions like `\quark_if_<quark>:NTF`, `\use_none_to_<quark>:wN`, and so on — this would make it much easier for 3rd party quark users to safely integrate quarks into their own code.

Not so sure about this one. Simply because I don't think this is a concept that should appear in the typical package code and so provide some more laborate mechanism that are not going to be used (or shouldn't be used normally)  seems overkill and gives the wrong message ... maybe.

**Comment by PhelypeOleinik** [comment]
(bringing this one up from the dead)

I was about to bug Bruno with an `unravel` issue when I found it to be a duplicate of this one:
```latex
\RequirePackage{unravel}
\ExplSyntaxOn
\def\infty{\if AB\q_nil\fi}
\unravel[prompt-input=u\q_nil]{\infty}
\stop
```

The `prompt-input` option stores its arguments in a `clist`, and the `\q_nil` trims the item (easier to see with `\clist_set:Nn \l_tmpa_clist { u\q_nil v }`).

I think the move to private quarks is a good thing. But having `\quark_new:N` define quark conditionals would be a little too much, in my opinion. However some way to define quark functions would be useful. Simple functions (with simple, I mean: they don't depend on a definition dedicated to one specific quark), like `\quark_if_recursion_tail_break:NN`, could have more general versions, like (say) `\quark_if_eq_break:NNN`.

However some conditionals like `\quark_if_recursion_tail_break:nN` require a `\__quark_if_recursion_tail:w`, which is defined in terms of the quark it's testing, for these perhaps something like `\quark_new_conditional:Nnn <quark> { <type> } { <signature> }` could be made. `<type>` would be `stop`, `stop_do`, `break`, etc. and the `<signature>` would figure out which version to define. The name of the resulting conditional would derive from the name of the quark and the other arguments... Thoughts?

**Comment by wspr** [comment]
Yes, it’s been a while but I remember the need to re-do quark-based commands from scratch if you weren’t using a kernel quark was … awkward.

On the other hand, how many people are defining their own quarks and using them extensively? I guess there are also advantages to keeping things simple.

**Comment by blefloch** [comment]
Yes, there should be some changes in quark usage in the kernel, but note
that this would not fix the issue with the unravel package noted above:
when unravelling kernel commands the new quarks would appear in the data
manipulated by unravel.  I need to make sure that unravel avoids any
kernel function that can fail upon encountering some quark (even
internal quarks).

**Comment by PhelypeOleinik** [comment]
Implemented in PR #690
